src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs

```diff
@@ -12,48 +12,44 @@
 namespace Microsoft.VisualStudio.LanguageServices.Experimentation
 {
     [Export(typeof(VisualStudioExperimentationService))]
-    [ExportWorkspaceService(typeof(IExperimentationService), ServiceLayer.Host), Shared]
-    internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedObject, IExperimentationService
+    [ExportWorkspaceService(typeof(IExperimentationServiceFactory), ServiceLayer.Host), Shared]
+    internal class VisualStudioExperimentationService : IExperimentationServiceFactory
     {
-        private readonly object _experimentationServiceOpt;
-        private readonly MethodInfo _isCachedFlightEnabledInfo;
-        private readonly IVsFeatureFlags _featureFlags;
+        private readonly IAsyncServiceProvider _serviceProvider;
+        private IExperimentationService _experimentationService;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public VisualStudioExperimentationService(IThreadingContext threadingContext, SVsServiceProvider serviceProvider)
-            : base(threadingContext)
+        public VisualStudioExperimentationService(SVsServiceProvider serviceProvider)
         {
-            object experimentationServiceOpt = null;
-            MethodInfo isCachedFlightEnabledInfo = null;
-            IVsFeatureFlags featureFlags = null;
+            _serviceProvider = (IAsyncServiceProvider)serviceProvider;
+        }
 
-            threadingContext.JoinableTaskFactory.Run(async () =>
+        public async ValueTask<IExperimentationService> GetExperimentationServiceAsync(CancellationToken cancellationToken)
+        {
+            if (_experimentationService is null)
             {
-                try
-                {
```

Comment:
my preference will be just kicking off task here. and then in IsExperimentEnabled, check whether services are set, otherwise, wait the task under JTF.Run rather than make everything async.

or if really want to make things async, just make IsExperiementEnabled Async.

Reply:
This is a more accurate representation of the asynchronous nature of the services involved. `IsExperimentEnabled` has no intrinsic asynchronous portion or cross-thread dependencies, but obtaining an instance of the service does.

Reply:
There's nothing wrong with making IExperimentationService async -- the remote implementation we checked in is doing this today:

https://github.com/dotnet/roslyn/blob/6ead316e86d304e8b5bf2b85b71cbdf35bb2b52c/src/Workspaces/Remote/Core/Services/RemoteExperimentationService.cs#L13-L17

We didn't make it async at the top level because of the uses in the constructor that you're fixing.

Reply:
> .Result

This is not a safe operation.